icu4c/source/test/intltest/localebuildertest.cpp

```diff
@@ -0,0 +1,1607 @@
+        Locale loc = bld.build(status);
+        if (('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') || ('0' <= ch && ch <= '9')) {
+            if (ch != 't' && ch != 'T' && ch != 'u' && ch != 'U' && ch != 'x' && ch != 'X') {
+                if (U_FAILURE(status)) {
+                    errln("setExtension('%c', \"%s\") should success but got Error: %s\n",
+                          ch, well_formed[ch % UPRV_LENGTHOF(well_formed)], u_errorName(status));
+                }
+            }
+        } else {
+            if (status != U_ILLEGAL_ARGUMENT_ERROR) {
+                errln("setExtension('%c', \"%s\") should fail but has no Error\n",
+                      ch, well_formed[ch % UPRV_LENGTHOF(well_formed)]);
+            }
+        }
+
+    }
+}
+
+void LocaleBuilderTest::TestSetExtensionValidateOthersIllFormed() {
+    static const char* ill_formed[] = {
+        "0",  // Too short
+        "a",  // Too short
+        "123456789",  // Too long
+        "abcdefghi",  // Too long
+        "ab-123456789",  // Second subtag too long
+        "abcdefghi-12",  // First subtag too long
+        "a-ab-87654321",  // Third subtag too long
+        "87654321-a-0-3",  // First subtag too long
+    };
+    char ch = 'a';
+    for (int i = 0; i < UPRV_LENGTHOF(ill_formed); i++, ch++) {
+        if (ch > 'z') ch = 'a';
+        UErrorCode status = U_ZERO_ERROR;
+        LocaleBuilder bld;
+        bld.setExtension(ch, ill_formed[i]);
+        Locale loc = bld.build(status);
+        if (status != U_ILLEGAL_ARGUMENT_ERROR) {
+            errln("setExtension('%c', \"%s\") should fail but has no Error\n",
+                  ch, ill_formed[i]);
+        }
+    }
+}
+
+void LocaleBuilderTest::TestSetLocale() {
+    LocaleBuilder bld1, bld2;
+    UErrorCode status = U_ZERO_ERROR;
+    Locale l1 = bld1.setLanguage("en")
+        .setScript("Latn")
+        .setRegion("MX")
+        .setVariant("3456-abcde")
+        .addUnicodeLocaleAttribute("456")
+        .addUnicodeLocaleAttribute("123")
+        .setUnicodeLocaleKeyword("nu", "thai")
+        .setUnicodeLocaleKeyword("co", "stroke")
+        .setUnicodeLocaleKeyword("ca", "chinese")
+        .build(status);
+    if (U_FAILURE(status) || l1.isBogus()) {
+        errln("build should success but got Error: %s\n", u_errorName(status));
+    }
+    status = U_ZERO_ERROR;
+    Locale l2 = bld1.setLocale(l1).build(status);
+    if (U_FAILURE(status) || l2.isBogus()) {
+        errln("build should success but got Error: %s\n", u_errorName(status));
+    }
+
+    if (l1 != l2) {
+        errln("Two locales should be the same, but one is '%s' and the other is '%s'",
+              l1.getName(), l2.getName());
+    }
+}
+
+void LocaleBuilderTest::TestPosixCases() {
+    UErrorCode status = U_ZERO_ERROR;
+    Locale l1 = Locale::forLanguageTag("en-US-u-va-posix", status);
+    if (U_FAILURE(status) || l1.isBogus()) {
+        errln("build should success but got Error: %s\n", u_errorName(status));
+    }
```

Comment:
Test that we should get the same as Locale("en-US-POSIX")? (Assuming we handle that?)